### ISSUE TITLE
Diagram explaining flow

### ISSUE THREAD

**Issue by mkg20001** [jacobheun/it-handshake#1]
Currently trying to make sense of this using diagrams.

(Isn't finished yet - `rest()` and `.stream` are missing)

I think it would be a good idea to have some kind of "diagram" like that one for every stream module, to get a better understanding of what these streams actually do

Thoughts?

![IMG_20191107_153259](https://user-images.githubusercontent.com/7735145/68397958-485b5280-0174-11ea-9529-24e9e83776d5.jpg)

![IMG_20191107_153310](https://user-images.githubusercontent.com/7735145/68397944-42657180-0174-11ea-983b-2a75aadc2bb1.jpg)

**Comment by jacobheun** [comment]
Yes, I think that would be helpful to better illustrate the intended flow. Also, after going through this if you have thoughts on changes to the api, method names, etc, it would be great to hear them. I think renaming some things list `rest` or simplifying what things are exposed could help make things easier to understand.

**Comment by mkg20001** [comment]
I think once there's a diagram, complexities will be much easier to _spot_ (pun intended :wink:)

(One api change that comes to my mind is to make stream a getter that automatically calls `.rest()`, but this might break passing the stream before calling `.rest()`, but unsure if that should even be supported)

Just one thing I can't get my head around: How should I display partially reading from a source and then re-using it in the diagram?

**Comment by jacobheun** [comment]
Does something like this help? My immediate thought went to a queue like diagram with multiple consumers of chunks from the queue.

![IMG_20191107_163623](https://user-images.githubusercontent.com/639834/68403305-04b91680-017d-11ea-9115-eeb293b5fec8.jpg)
